### DISCUSSION TITLE
Zustand persist with AsyncStorage

### DISCUSSION THREAD

**Discussion by alicanso** [pmndrs/zustand#1567] [Q&A] [answered]
## Summary

using storage instead of getStorage causes error.

here is my code:

```
import { createJSONStorage, persist } from 'zustand/middleware'
import { Info } from '../types/info'
import { Store } from '../types/store'
import AsyncStorage from '@react-native-async-storage/async-storage'

const useMainStore = create<Store>()(
    persist(
        (set, get) => ({
            data: '',
        }),
        {
            name: 'main-storage',
            storage: createJSONStorage(() => getStorageType())
        }
    )
)

export const getStorageType = () => {
    const isBrowser = typeof window !== 'undefined' //browser or react-native
    return !isBrowser ? AsyncStorage : window.localStorage
}
```

`the error is: TypeError: Cannot read property 'setItem' of undefined`
<img width="787" alt="image" src="https://user-images.githubusercontent.com/67914872/214003417-bc0015d1-a874-4df8-9ec1-068ac475b14a.png">

## Link to reproduction

just change getStorage with storage as described in documentation.

**Comment by dai-shi** [comment]
What's the error message?
Does it cause error in browser too?

**Reply by alicanso** [reply]
> What's the error message? Does it cause error in browser too?

I am working on react-native.

**Reply by dai-shi** [reply]
So, does this cause the error? Can you check what methods `AsyncStorage` has?
```ts
import { create } from 'zsutand'
import { createJSONStorage, persist } from 'zustand/middleware'
import { Store } from '../types/store'
import AsyncStorage from '@react-native-async-storage/async-storage'

const useMainStore = create<Store>()(
    persist(
        (set, get) => ({
            data: '',
        }),
        {
            name: 'main-storage',
            storage: createJSONStorage(() => AsyncStorage)
        }
    )
)
```

**Reply by alicanso** [reply]
> So, does this cause the error? Can you check what methods `AsyncStorage` has?
> 
> ```ts
> import { create } from 'zsutand'
> import { createJSONStorage, persist } from 'zustand/middleware'
> import { Store } from '../types/store'
> import AsyncStorage from '@react-native-async-storage/async-storage'
> 
> const useMainStore = create<Store>()(
>     persist(
>         (set, get) => ({
>             data: '',
>         }),
>         {
>             name: 'main-storage',
>             storage: createJSONStorage(() => AsyncStorage)
>         }
>     )
> )
> ```

I checked this before. It has setItem method.

<img width="734" alt="image" src="https://user-images.githubusercontent.com/67914872/214015526-5c83fe68-9be5-4210-b4a6-892d8f09e362.png">

**Reply by dai-shi** [reply]
Thanks for confirming.
Hm, I'm not sure what's wrong.
I hope someone who has experience with RN can help this issue.

**Comment by alicanso** [comment, marked as answer]
I found the problem. The problem is **window** object is **not** undefined in expo environment. So my _getStorageType_ function returns localstorage to use. I changed **window** with **document** and problem has been solved!
